### DISCUSSION TITLE
connectOrCreate multiple records issue

### DISCUSSION THREAD

**Discussion by hmaiermack** [prisma/prisma#12059] [Q&A] [answered]
Having an issue with connectOrCreate accepting my array that I'm pretty sure is properly structured.

Models are as follows
```prisma
model Task {
  id Int @id @default(autoincrement())

  title     String
  timeStart DateTime
  timeEnd   DateTime

  user   User @relation(fields: [userId], references: [id])
  userId Int

  tags TagsOnTasks[]
}

model Tag {
  id Int @id @default(autoincrement())

  name String @unique

  colorHexValue String @default("#5F5F5F")

  tasks TagsOnTasks[]

}

model TagsOnTasks {
  task   Task @relation(fields: [taskId], references: [id])
  taskId Int

  tag   Tag @relation(fields: [tagId], references: [id])
  tagId Int

  @@id([taskId, tagId])
}
```

Service layer
```TypeScript
    /*    this map results in something roughly like this
    [
        {
            where: { name: 'tag1' },
            create: { name: 'tag1', colorHexValue: '#FFF' },
        },
        {
            where: { name: 'tag2' },
            create: { name: 'tag2', colorHexValue: '#FFF' },
        }
    ]
    */

    const connectOrCreateArrayForTags = tags.map((tag) =>
      tag.colorHexValue
        ? {
            where: { name: tag.name },
            create: { name: tag.name, colorHexValue: tag.colorHexValue },
          }
        : { where: { name: tag.name }, create: { name: tag.name } },
    );

    const task = await this.prisma.task.create({
      data: {
        title: dto.title,
        timeStart: dto.timeStart,
        timeEnd: dto.timeEnd,
        userId,
        tags: {
          connectOrCreate: connectOrCreateArrayForTags,
        },
      },
    });

```
And the error
```TypeScript
(property) connectOrCreate: {
    where: {
        name: string;
    };
    create: {
        name: string;
        colorHexValue: string;
    };
}[]
Type '{ where: { name: string; }; create: { name: string; colorHexValue: string; }; }[]' is not assignable to type 'Enumerable<TagsOnTasksCreateOrConnectWithoutTaskInput>'.
  Type '{ where: { name: string; }; create: { name: string; colorHexValue: string; }; }[]' is not assignable to type 'TagsOnTasksCreateOrConnectWithoutTaskInput[]'.
    Type '{ where: { name: string; }; create: { name: string; colorHexValue: string; }; }' is not assignable to type 'TagsOnTasksCreateOrConnectWithoutTaskInput'.
      Types of property 'where' are incompatible.
        Type '{ name: string; }' has no properties in common with type 'TagsOnTasksWhereUniqueInput'.ts(2322)
[index.d.ts(5361, 5): ]()The expected type comes from property 'connectOrCreate' which is declared here on type 'TagsOnTasksUncheckedCreateNestedManyWithoutTaskInput | TagsOnTasksCreateNestedManyWithoutTaskInput'
```

I'm still relatively new to TS and haven't been able to figure out the problem from the error message. Possibly related to the colorHexValue ternary in the map in the service layer? It's late so I could be missing something obvious, would love some help!

**Comment by austincrim** [comment]
Hey @hmaiermack 👋 , thanks for the question!

The issue here is that `tasks.tags` is actually of type `TagsOnTasks`, not `Tags`. This means the `where` property of your `connectOrCreate` should look like this to satisfy the `TagsOnTasksWhereUniqueInput` type in the error message:
```ts
...
connectOrCreate: {
  where: {
    taskId_tagId: {
      taskId: someId,
      tagId: anotherId
    }
  }
}
...
```

This is because your schema specifies `@@id([taskId, tagId])` for the `TagsOnTasks` join table.

Let me know if I can clarify anything!

**Comment by hmaiermack** [comment]
@austincrim thanks for the response! If I'm understanding correctly, does this mean that you're unable to use `connectOrCreate` on an explicit many to many relationship to create the new record before creating the JOIN table as it need id's from both records (meaning each needs to already exist in the DB)? Or would it just involve another level of nesting the query?

I dropped the explicit JOIN table and now it is working as expected, just curious! Thanks!

**Reply by psibean** [reply, marked as answer]
I believe the explicit join does mean you can't do it this way.

Instead you'd have to use the join table create method, and do the nested create for each relation under it, where you'd then have an array of tags.

**Comment by aruns05** [comment]
I am facing similar issue , In the table1 object_id is a unique autoincrement field. When try to run this get the below error

        table1:{
          connectOrCreate: obj.hubDetails.map(s =>({
            create:{
              field1: s.field1,
              field2:s.field2,
              field3: s.field3,
         
            }
          }))
  
        }

  },

**error**
Argument where for data.table1.connectOrCreate.0.where is missing.
        +         where: {
        +           object_id?: Decimal
        +         }
                }

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
